One last issue I noticed in Travis, after this it should be all fine.
```bash
./src/stc/PlatWX.cpp:2722:11: warning: private field 'm_surfaceFontData' is not used [-Wunused-private-field]
    void* m_surfaceFontData;
          ^
1 warning generated.
```

--- a/src/stc/PlatWX.cpp
+++ b/src/stc/PlatWX.cpp
@@ -2651,8 +2651,7 @@ int wxSTCListBoxVisualData::GetStartLen() const
 class wxSTCListBox : public wxSystemThemedControl<wxVListBox>
 {
 public:
-    wxSTCListBox(wxWindow*, wxSTCListBoxVisualData*, int, int);
-    ~wxSTCListBox();
+    wxSTCListBox(wxWindow*, wxSTCListBoxVisualData*, int);
 
     // wxWindow overrides
     virtual bool AcceptsFocus() const wxOVERRIDE;
@@ -2662,7 +2661,7 @@ public:
     void SetContainerBorderSize(int);
 
     // ListBoxImpl implementation
-    void SetListBoxFont(Font &font);
+    virtual void SetListBoxFont(Font &font);
     void SetAverageCharWidth(int width);
     PRectangle GetDesiredRect() const;
     int CaretFromEdge() const;
@@ -2694,7 +2693,6 @@ protected:
     virtual void OnDrawItem(wxDC& , const wxRect &, size_t) const wxOVERRIDE;
     virtual void OnDrawBackground(wxDC&, const wxRect&,size_t) const wxOVERRIDE;
 
-private:
     wxSTCListBoxVisualData* m_visualData;
     wxVector<wxString>      m_labels;
     wxVector<int>           m_imageNos;
@@ -2716,21 +2714,14 @@ private:
     int m_imagePadding;
     int m_textBoxToTextGap;
     int m_textExtraVerticalPadding;
-
-    // These are needed when Direct2D is used for drawing.
-    int m_technology;
-#ifdef HAVE_DIRECTWRITE_TECHNOLOGY
-    SurfaceFontDataD2D* m_surfaceFontData;
-#endif
 };
 
-wxSTCListBox::wxSTCListBox(wxWindow* parent, wxSTCListBoxVisualData* v,
-                           int ht, int technology)
+wxSTCListBox::wxSTCListBox(wxWindow* parent, wxSTCListBoxVisualData* v, int ht)
              :wxSystemThemedControl<wxVListBox>(),
               m_visualData(v), m_maxStrWidth(0), m_currentRow(wxNOT_FOUND),
               m_doubleClickAction(NULL), m_doubleClickActionData(NULL),
               m_aveCharWidth(8), m_textHeight(ht), m_itemHeight(ht),
-              m_textTopGap(0), m_technology(technology)
+              m_textTopGap(0)
 {
     wxVListBox::Create(parent, wxID_ANY, wxDefaultPosition, wxDefaultSize,
                        wxBORDER_NONE, "AutoCompListBox");
@@ -2759,20 +2750,6 @@ wxSTCListBox::wxSTCListBox(wxWindow* parent, wxSTCListBoxVisualData* v,
             parent->SetOwnBackgroundColour(m_visualData->GetBgColour());
         #endif
     }
-
-    #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
-        m_surfaceFontData = NULL;
-    #endif
-}
-
-wxSTCListBox::~wxSTCListBox()
-{
-#ifdef HAVE_DIRECTWRITE_TECHNOLOGY
-    if ( m_surfaceFontData )
-    {
-        delete m_surfaceFontData;
-    }
-#endif
 }
 
 bool wxSTCListBox::AcceptsFocus() const
@@ -2792,29 +2769,9 @@ void wxSTCListBox::SetContainerBorderSize(int s)
 
 void wxSTCListBox::SetListBoxFont(Font &font)
 {
-    if ( m_technology == wxSTC_TECHNOLOGY_DIRECTWRITE )
-    {
-    #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
-        wxFontWithAscent* fwa = wxFontWithAscent::FromFID(font.GetID());
-
-        SurfaceData* data = fwa->GetSurfaceFontData();
-        SurfaceFontDataD2D* d2dft = static_cast<SurfaceFontDataD2D*>(data);
-        m_surfaceFontData = new SurfaceFontDataD2D(*d2dft);
-
-        SurfaceD2D surface;
-        wxClientDC dc(this);
-        surface.Init(&dc,GetGrandParent());
-        m_textHeight = surface.Height(font);
-        surface.Release();
-    #endif
-    }
-    else
-    {
-        SetFont(*((wxFont*)font.GetID()));
-        int w;
-        GetTextExtent(EXTENT_TEST, &w, &m_textHeight);
-    }
-
+    SetFont(*((wxFont*)font.GetID()));
+    int w;
+    GetTextExtent(EXTENT_TEST, &w, &m_textHeight);
     RecalculateItemHeight();
 }
 
@@ -3084,69 +3041,11 @@ void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
     else
         textCol = m_visualData->GetTextColour();
 
-    if ( m_technology == wxSTC_TECHNOLOGY_DIRECTWRITE )
-    {
-    #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
-        SurfaceD2D surface;
-        surface.Init(&dc, GetGrandParent());
+    wxDCTextColourChanger tcc(dc, textCol);
 
-        wxString ellipsizedLabel = label;
-
-        wxFontWithAscent* fontCopy = new wxFontWithAscent(wxFont());
-        SurfaceFontDataD2D* newSurfaceData =
-            new SurfaceFontDataD2D(*m_surfaceFontData);
-        fontCopy->SetSurfaceFontData(newSurfaceData);
-        Font tempFont;
-        tempFont.SetID(fontCopy);
-
-        wxCharBuffer buffer = wx2stc(label);
-        int curWidth = surface.WidthText(tempFont, buffer.data(),
-                                          wx2stclen(wxString(),buffer));
-
-        int maxWidth = rect.GetWidth() - leftGap;
-
-        if ( curWidth > maxWidth )
-        {
-            int len = label.Length();
-
-            for ( int i = len ; i >= 0 ; --i)
-            {
-                ellipsizedLabel = label.Left(i);
-                ellipsizedLabel << "...";
-
-                buffer = wx2stc(ellipsizedLabel);
-                curWidth = surface.WidthText(tempFont, buffer.data(),
-                                             wx2stclen(wxString(),buffer));
-
-                if ( curWidth <= maxWidth )
-                {
-                    break;
-                }
-            }
-        }
-
-        wxRect rect2(rect.GetLeft() + leftGap, rect.GetTop() + topGap,
-                    rect.GetWidth() - leftGap, m_textHeight);
-
-        PRectangle prect = PRectangleFromwxRect(rect2);
-        ColourDesired fore(textCol.Red(), textCol.Green(), textCol.Blue());
-
-        XYPOSITION ybase = rect2.GetTop() + m_surfaceFontData->GetAscent();
-
-        surface.DrawTextTransparent(prect, tempFont, ybase, buffer.data(),
-                                    wx2stclen(wxString(),buffer), fore);
-        tempFont.Release();
-        surface.Release();
-    #endif // HAVE_DIRECTWRITE_TECHNOLOGY
-    }
-    else
-    {
-        wxDCTextColourChanger tcc(dc, textCol);
-
-        label = wxControl::Ellipsize(label, dc, wxELLIPSIZE_END,
-                                     rect.GetWidth() - leftGap);
-        dc.DrawText(label, rect.GetLeft() + leftGap, rect.GetTop() + topGap);
-    }
+    label = wxControl::Ellipsize(label, dc, wxELLIPSIZE_END,
+                                 rect.GetWidth() - leftGap);
+    dc.DrawText(label, rect.GetLeft() + leftGap, rect.GetTop() + topGap);
 
     const wxBitmap* b = m_visualData->GetImage(imageNo);
     if ( b )
@@ -3214,6 +3113,130 @@ void wxSTCListBox::OnDrawBackground(wxDC &dc, const wxRect &rect,size_t n) const
 }
 
 
+#ifdef HAVE_DIRECTWRITE_TECHNOLOGY
+class wxSTCListBoxD2D : public wxSTCListBox
+{
+public:
+    wxSTCListBoxD2D(wxWindow* parent, wxSTCListBoxVisualData* v, int ht)
+        : wxSTCListBox(parent, v, ht)
+        , m_surfaceFontData(NULL)
+    {
+    }
+
+    ~wxSTCListBoxD2D()
+    {
+        if ( m_surfaceFontData )
+        {
+            delete m_surfaceFontData;
+        }
+    }
+
+    void SetListBoxFont(Font& font) wxOVERRIDE
+    {
+        wxFontWithAscent* fwa = wxFontWithAscent::FromFID(font.GetID());
+
+        SurfaceData* data = fwa->GetSurfaceFontData();
+        SurfaceFontDataD2D* d2dft = static_cast<SurfaceFontDataD2D*>(data);
+        m_surfaceFontData = new SurfaceFontDataD2D(*d2dft);
+
+        SurfaceD2D surface;
+        wxClientDC dc(this);
+        surface.Init(&dc, GetGrandParent());
+        m_textHeight = surface.Height(font);
+        surface.Release();
+
+        RecalculateItemHeight();
+    }
+
+    void OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const wxOVERRIDE
+    {
+        wxString label;
+        int imageNo = -1;
+        if ( n < m_labels.size() )
+        {
+            label = m_labels[n];
+            imageNo = m_imageNos[n];
+        }
+
+        int topGap = m_textTopGap;
+        int leftGap = TextBoxFromClientEdge() + m_textBoxToTextGap;
+
+        wxColour textCol;
+
+        if ( IsSelected(n) )
+            textCol = m_visualData->GetHighlightTextColour();
+        else if ( static_cast<int>(n) == m_currentRow )
+            textCol = m_visualData->GetCurrentTextColour();
+        else
+            textCol = m_visualData->GetTextColour();
+
+        SurfaceD2D surface;
+        surface.Init(&dc, GetGrandParent());
+
+        wxString ellipsizedLabel = label;
+
+        wxFontWithAscent* fontCopy = new wxFontWithAscent(wxFont());
+        SurfaceFontDataD2D* newSurfaceData =
+            new SurfaceFontDataD2D(*m_surfaceFontData);
+        fontCopy->SetSurfaceFontData(newSurfaceData);
+        Font tempFont;
+        tempFont.SetID(fontCopy);
+
+        wxCharBuffer buffer = wx2stc(label);
+        int curWidth = surface.WidthText(tempFont, buffer.data(),
+                                         wx2stclen(wxString(), buffer));
+
+        int maxWidth = rect.GetWidth() - leftGap;
+
+        if ( curWidth > maxWidth )
+        {
+            int len = label.Length();
+
+            for ( int i = len; i >= 0; --i )
+            {
+                ellipsizedLabel = label.Left(i);
+                ellipsizedLabel << "...";
+
+                buffer = wx2stc(ellipsizedLabel);
+                curWidth = surface.WidthText(tempFont, buffer.data(),
+                                              wx2stclen(wxString(), buffer));
+
+                if ( curWidth <= maxWidth )
+                {
+                    break;
+                }
+            }
+        }
+
+        wxRect rect2(rect.GetLeft() + leftGap, rect.GetTop() + topGap,
+                     rect.GetWidth() - leftGap, m_textHeight);
+
+        PRectangle prect = PRectangleFromwxRect(rect2);
+        ColourDesired fore(textCol.Red(), textCol.Green(), textCol.Blue());
+
+        XYPOSITION ybase = rect2.GetTop() + m_surfaceFontData->GetAscent();
+
+        surface.DrawTextTransparent(prect, tempFont, ybase, buffer.data(),
+                                    wx2stclen(wxString(), buffer), fore);
+        tempFont.Release();
+        surface.Release();
+
+        const wxBitmap* b = m_visualData->GetImage(imageNo);
+        if ( b )
+        {
+            const int width = m_visualData->GetImageAreaWidth();
+            topGap = (m_itemHeight - b->GetHeight()) / 2;
+            leftGap = m_imagePadding + (width - b->GetWidth()) / 2;
+            dc.DrawBitmap(*b, rect.GetLeft() + leftGap, rect.GetTop() + topGap, true);
+        }
+    }
+
+private:
+    SurfaceFontDataD2D* m_surfaceFontData;
+};
+#endif
+
+
 // A popup window to place the wxSTCListBox upon
 class wxSTCListBoxWin : public wxSTCPopupWindow
 {
@@ -3232,7 +3255,16 @@ wxSTCListBoxWin::wxSTCListBoxWin(wxWindow* parent, wxSTCListBox** lb,
                                  wxSTCListBoxVisualData* v, int h, int tech)
                 :wxSTCPopupWindow(parent)
 {
-    *lb = new wxSTCListBox(this, v, h, tech);
+#ifdef HAVE_DIRECTWRITE_TECHNOLOGY
+    if ( tech == wxSTC_TECHNOLOGY_DIRECTWRITE )
+    {
+        *lb = new wxSTCListBoxD2D(this, v, h);
+    }
+    else
+#endif
+    {
+        *lb = new wxSTCListBox(this, v, h);
+    }
 
     // Use the background of this window to form a frame around the listbox
     // except on macos where the native Scintilla popup has no frame.

--- a/src/stc/PlatWX.cpp
+++ b/src/stc/PlatWX.cpp
@@ -2679,6 +2679,8 @@ protected:
     void SelectHelper(int i);
     void RecalculateItemHeight();
     int TextBoxFromClientEdge() const;
+    virtual void OnDrawItemText(wxDC&, const wxRect&,
+                                const wxString&, const wxColour&) const;
 
     // Event handlers
     void OnSelection(wxCommandEvent&);
@@ -3019,6 +3021,17 @@ wxCoord wxSTCListBox::OnMeasureItem(size_t WXUNUSED(n)) const
 //       x = m_imagePadding + m_imageAreaWidth + m_imagePadding.
 // Text is drawn at x + m_textBoxToTextGap and centered vertically.
 
+void wxSTCListBox::OnDrawItemText(wxDC& dc, const wxRect& rect,
+                                  const wxString& label,
+                                  const wxColour& textCol) const
+{
+    wxDCTextColourChanger tcc(dc, textCol);
+
+    wxString ellipsizedlabel = wxControl::Ellipsize(label, dc, wxELLIPSIZE_END,
+                                                    rect.GetWidth());
+    dc.DrawText(ellipsizedlabel, rect.GetLeft(), rect.GetTop());
+}
+
 void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
 {
     wxString label;
@@ -3041,11 +3054,10 @@ void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
     else
         textCol = m_visualData->GetTextColour();
 
-    wxDCTextColourChanger tcc(dc, textCol);
+    wxRect rect2(rect.GetLeft() + leftGap, rect.GetTop() + topGap,
+                 rect.GetWidth() - leftGap, m_textHeight);
 
-    label = wxControl::Ellipsize(label, dc, wxELLIPSIZE_END,
-                                 rect.GetWidth() - leftGap);
-    dc.DrawText(label, rect.GetLeft() + leftGap, rect.GetTop() + topGap);
+    OnDrawItemText(dc, rect2, label, textCol);
 
     const wxBitmap* b = m_visualData->GetImage(imageNo);
     if ( b )
@@ -3148,47 +3160,25 @@ public:
         RecalculateItemHeight();
     }
 
-    void OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const wxOVERRIDE
+    void OnDrawItemText(wxDC& dc, const wxRect& rect, const wxString& label,
+                        const wxColour& textCol) const wxOVERRIDE
     {
-        wxString label;
-        int imageNo = -1;
-        if ( n < m_labels.size() )
-        {
-            label = m_labels[n];
-            imageNo = m_imageNos[n];
-        }
-
-        int topGap = m_textTopGap;
-        int leftGap = TextBoxFromClientEdge() + m_textBoxToTextGap;
-
-        wxColour textCol;
-
-        if ( IsSelected(n) )
-            textCol = m_visualData->GetHighlightTextColour();
-        else if ( static_cast<int>(n) == m_currentRow )
-            textCol = m_visualData->GetCurrentTextColour();
-        else
-            textCol = m_visualData->GetTextColour();
+        wxFontWithAscent* fontCopy = new wxFontWithAscent(wxFont());
+        SurfaceFontDataD2D* sfd = new SurfaceFontDataD2D(*m_surfaceFontData);
+        fontCopy->SetSurfaceFontData(sfd);
+        Font tempFont;
+        tempFont.SetID(fontCopy);
 
         SurfaceD2D surface;
         surface.Init(&dc, GetGrandParent());
 
         wxString ellipsizedLabel = label;
 
-        wxFontWithAscent* fontCopy = new wxFontWithAscent(wxFont());
-        SurfaceFontDataD2D* newSurfaceData =
-            new SurfaceFontDataD2D(*m_surfaceFontData);
-        fontCopy->SetSurfaceFontData(newSurfaceData);
-        Font tempFont;
-        tempFont.SetID(fontCopy);
-
-        wxCharBuffer buffer = wx2stc(label);
+        wxCharBuffer buffer = wx2stc(ellipsizedLabel);
         int curWidth = surface.WidthText(tempFont, buffer.data(),
                                          wx2stclen(wxString(), buffer));
 
-        int maxWidth = rect.GetWidth() - leftGap;
-
-        if ( curWidth > maxWidth )
+        if ( curWidth > rect.GetWidth() )
         {
             int len = label.Length();
 
@@ -3201,34 +3191,22 @@ public:
                 curWidth = surface.WidthText(tempFont, buffer.data(),
                                               wx2stclen(wxString(), buffer));
 
-                if ( curWidth <= maxWidth )
+                if ( curWidth <= rect.GetWidth() )
                 {
                     break;
                 }
             }
         }
 
-        wxRect rect2(rect.GetLeft() + leftGap, rect.GetTop() + topGap,
-                     rect.GetWidth() - leftGap, m_textHeight);
-
-        PRectangle prect = PRectangleFromwxRect(rect2);
+        PRectangle prect = PRectangleFromwxRect(rect);
         ColourDesired fore(textCol.Red(), textCol.Green(), textCol.Blue());
 
-        XYPOSITION ybase = rect2.GetTop() + m_surfaceFontData->GetAscent();
+        XYPOSITION ybase = rect.GetTop() + m_surfaceFontData->GetAscent();
 
         surface.DrawTextTransparent(prect, tempFont, ybase, buffer.data(),
                                     wx2stclen(wxString(), buffer), fore);
         tempFont.Release();
         surface.Release();
-
-        const wxBitmap* b = m_visualData->GetImage(imageNo);
-        if ( b )
-        {
-            const int width = m_visualData->GetImageAreaWidth();
-            topGap = (m_itemHeight - b->GetHeight()) / 2;
-            leftGap = m_imagePadding + (width - b->GetWidth()) / 2;
-            dc.DrawBitmap(*b, rect.GetLeft() + leftGap, rect.GetTop() + topGap, true);
-        }
     }
 
 private:

--- a/src/stc/PlatWX.cpp
+++ b/src/stc/PlatWX.cpp
@@ -2360,8 +2360,7 @@ private:
 
     int      m_desiredVisibleRows;
     ImgList  m_imgList;
-    int      m_imageAreaWidth;
-    int      m_imageAreaHeight;
+    wxSize   m_imgAreaSize;
 
     wxColour m_borderColour;
     wxColour m_bgColour;
@@ -2385,8 +2384,7 @@ private:
 };
 
 wxSTCListBoxVisualData::wxSTCListBoxVisualData(int d):m_desiredVisibleRows(d),
-                        m_imageAreaWidth(0), m_imageAreaHeight(0),
-                        m_useDefaultBgColour(true),
+                        m_imgAreaSize(0,0), m_useDefaultBgColour(true),
                         m_useDefaultTextColour(true),
                         m_useDefaultHighlightBgColour(true),
                         m_useDefaultHighlightTextColour(true),
@@ -2419,39 +2417,32 @@ void wxSTCListBoxVisualData::RegisterImage(int type, const wxBitmap& bmp)
         return;
 
     ImgList::iterator it=m_imgList.find(type);
-    bool preExisting = false;
+    bool preExistingWithDifferentSize = false;
     if ( it != m_imgList.end() )
     {
+        if ( it->second.GetSize() != bmp.GetSize() )
+        {
+            preExistingWithDifferentSize = true;
+        }
+
         m_imgList.erase(it);
-        preExisting = true;
     }
 
     m_imgList[type] = bmp;
 
-    if ( preExisting )
+    if ( preExistingWithDifferentSize )
     {
-        m_imageAreaWidth = 0;
-        m_imageAreaHeight = 0;
+        m_imgAreaSize.Set(0,0);
 
-        for ( ImgList::iterator it = m_imgList.begin() ; it != m_imgList.end() ; ++it )
+        for ( ImgList::iterator imgIt = m_imgList.begin() ;
+              imgIt != m_imgList.end() ; ++imgIt )
         {
-            wxBitmap bmp = it->second;
-
-            if ( bmp.GetWidth() > m_imageAreaWidth )
-            {
-                m_imageAreaWidth = bmp.GetWidth();
-            }
-
-            if ( bmp.GetHeight() > m_imageAreaHeight )
-            {
-                m_imageAreaHeight = bmp.GetHeight();
-            }
+            m_imgAreaSize.IncTo(it->second.GetSize());
         }
     }
     else
     {
-        m_imageAreaWidth = wxMax(m_imageAreaWidth, bmp.GetWidth());
-        m_imageAreaHeight = wxMax(m_imageAreaHeight, bmp.GetHeight());
+        m_imgAreaSize.IncTo(bmp.GetSize());
     }
 }
 
@@ -2486,8 +2477,7 @@ void wxSTCListBoxVisualData::RegisterRGBAImage(int type, int width, int height,
 void wxSTCListBoxVisualData::ClearRegisteredImages()
 {
     m_imgList.clear();
-    m_imageAreaWidth = 0;
-    m_imageAreaHeight = 0;
+    m_imgAreaSize.Set(0,0);
 }
 
 const wxBitmap* wxSTCListBoxVisualData::GetImage(int i) const
@@ -2502,12 +2492,12 @@ const wxBitmap* wxSTCListBoxVisualData::GetImage(int i) const
 
 int wxSTCListBoxVisualData::GetImageAreaWidth() const
 {
-    return m_imageAreaWidth;
+    return m_imgAreaSize.GetWidth();
 }
 
 int wxSTCListBoxVisualData::GetImageAreaHeight() const
 {
-    return m_imageAreaHeight;
+    return m_imgAreaSize.GetHeight();
 }
 
 void wxSTCListBoxVisualData::ComputeColours()
@@ -2729,7 +2719,9 @@ private:
 
     // These are needed when Direct2D is used for drawing.
     int m_technology;
-    void* m_surfaceFontData;
+#ifdef HAVE_DIRECTWRITE_TECHNOLOGY
+    SurfaceFontDataD2D* m_surfaceFontData;
+#endif
 };
 
 wxSTCListBox::wxSTCListBox(wxWindow* parent, wxSTCListBoxVisualData* v,
@@ -2738,7 +2730,7 @@ wxSTCListBox::wxSTCListBox(wxWindow* parent, wxSTCListBoxVisualData* v,
               m_visualData(v), m_maxStrWidth(0), m_currentRow(wxNOT_FOUND),
               m_doubleClickAction(NULL), m_doubleClickActionData(NULL),
               m_aveCharWidth(8), m_textHeight(ht), m_itemHeight(ht),
-              m_textTopGap(0), m_technology(technology), m_surfaceFontData(NULL)
+              m_textTopGap(0), m_technology(technology)
 {
     wxVListBox::Create(parent, wxID_ANY, wxDefaultPosition, wxDefaultSize,
                        wxBORDER_NONE, "AutoCompListBox");
@@ -2767,6 +2759,10 @@ wxSTCListBox::wxSTCListBox(wxWindow* parent, wxSTCListBoxVisualData* v,
             parent->SetOwnBackgroundColour(m_visualData->GetBgColour());
         #endif
     }
+
+    #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
+        m_surfaceFontData = NULL;
+    #endif
 }
 
 wxSTCListBox::~wxSTCListBox()
@@ -2774,9 +2770,7 @@ wxSTCListBox::~wxSTCListBox()
 #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
     if ( m_surfaceFontData )
     {
-        SurfaceFontDataD2D* fontData =
-            reinterpret_cast<SurfaceFontDataD2D*>(m_surfaceFontData);
-        delete fontData;
+        delete m_surfaceFontData;
     }
 #endif
 }
@@ -2798,13 +2792,7 @@ void wxSTCListBox::SetContainerBorderSize(int s)
 
 void wxSTCListBox::SetListBoxFont(Font &font)
 {
-    if ( m_technology == wxSTC_TECHNOLOGY_DEFAULT )
-    {
-        SetFont(*((wxFont*)font.GetID()));
-        int w;
-        GetTextExtent(EXTENT_TEST, &w, &m_textHeight);
-    }
-    else
+    if ( m_technology == wxSTC_TECHNOLOGY_DIRECTWRITE )
     {
     #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
         wxFontWithAscent* fwa = wxFontWithAscent::FromFID(font.GetID());
@@ -2813,19 +2801,18 @@ void wxSTCListBox::SetListBoxFont(Font &font)
         SurfaceFontDataD2D* d2dft = static_cast<SurfaceFontDataD2D*>(data);
         m_surfaceFontData = new SurfaceFontDataD2D(*d2dft);
 
-        wxFontWithAscent* fontCopy = new wxFontWithAscent(wxFont());
-        SurfaceFontDataD2D* newSurfaceData = new SurfaceFontDataD2D(*d2dft);
-        fontCopy->SetSurfaceFontData(newSurfaceData);
-        Font tempFont;
-        tempFont.SetID(fontCopy);
-
         SurfaceD2D surface;
         wxClientDC dc(this);
         surface.Init(&dc,GetGrandParent());
-        m_textHeight = surface.Height(tempFont);
-        tempFont.Release();
+        m_textHeight = surface.Height(font);
         surface.Release();
     #endif
+    }
+    else
+    {
+        SetFont(*((wxFont*)font.GetID()));
+        int w;
+        GetTextExtent(EXTENT_TEST, &w, &m_textHeight);
     }
 
     RecalculateItemHeight();
@@ -3097,15 +3084,7 @@ void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
     else
         textCol = m_visualData->GetTextColour();
 
-    if ( m_technology == wxSTC_TECHNOLOGY_DEFAULT )
-    {
-        wxDCTextColourChanger tcc(dc, textCol);
-
-        label = wxControl::Ellipsize(label, dc, wxELLIPSIZE_END,
-                                     rect.GetWidth() - leftGap);
-        dc.DrawText(label, rect.GetLeft() + leftGap, rect.GetTop() + topGap);
-    }
-    else
+    if ( m_technology == wxSTC_TECHNOLOGY_DIRECTWRITE )
     {
     #ifdef HAVE_DIRECTWRITE_TECHNOLOGY
         SurfaceD2D surface;
@@ -3113,10 +3092,9 @@ void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
 
         wxString ellipsizedLabel = label;
 
-        SurfaceFontDataD2D* fontData =
-            reinterpret_cast<SurfaceFontDataD2D*>(m_surfaceFontData);
         wxFontWithAscent* fontCopy = new wxFontWithAscent(wxFont());
-        SurfaceFontDataD2D* newSurfaceData = new SurfaceFontDataD2D(*fontData);
+        SurfaceFontDataD2D* newSurfaceData =
+            new SurfaceFontDataD2D(*m_surfaceFontData);
         fontCopy->SetSurfaceFontData(newSurfaceData);
         Font tempFont;
         tempFont.SetID(fontCopy);
@@ -3153,13 +3131,21 @@ void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
         PRectangle prect = PRectangleFromwxRect(rect2);
         ColourDesired fore(textCol.Red(), textCol.Green(), textCol.Blue());
 
-        XYPOSITION ybase = rect2.GetTop() + fontData->GetAscent();
+        XYPOSITION ybase = rect2.GetTop() + m_surfaceFontData->GetAscent();
 
         surface.DrawTextTransparent(prect, tempFont, ybase, buffer.data(),
                                     wx2stclen(wxString(),buffer), fore);
         tempFont.Release();
         surface.Release();
     #endif // HAVE_DIRECTWRITE_TECHNOLOGY
+    }
+    else
+    {
+        wxDCTextColourChanger tcc(dc, textCol);
+
+        label = wxControl::Ellipsize(label, dc, wxELLIPSIZE_END,
+                                     rect.GetWidth() - leftGap);
+        dc.DrawText(label, rect.GetLeft() + leftGap, rect.GetTop() + topGap);
     }
 
     const wxBitmap* b = m_visualData->GetImage(imageNo);

